identity/src/keypair.rs

```diff
@@ -342,6 +342,59 @@ impl Keypair {
             KeyPairInner::Ecdsa(_) => KeyType::Ecdsa,
         }
     }
+
+    /// Deterministically derive a new secret from this [`Keypair`], taking into account the provided domain.
+    ///
+    /// This works for all key types except RSA where it returns `None`.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// # fn main() {
+    /// # use libp2p_identity as identity;
+    ///
+    /// let key = identity::Keypair::generate_ed25519();
+    ///
+    /// let new_key = key.derive_secret(b"my encryption key").expect("can derive secret for ed25519");
+    /// # }
+    /// ```
+    #[allow(unused_variables, unreachable_code)]
+    pub fn derive_secret(&self, domain: &[u8]) -> Option<[u8; 32]> {
+        #[cfg(any(
+            feature = "ecdsa",
+            feature = "secp256k1",
+            feature = "ed25519",
+            feature = "rsa"
+        ))]
+        return Some(
+            hkdf::Hkdf::<sha2::Sha256>::extract(None, &[domain, &self.secret()?].concat())
```

Comment:
Follow-up question for an hkdf novice:

Do I understand correctly, that we don't need the _expand_ step here, as a `[u8; 32]` is large enough for all use-cases and as we only need a single key per `domain`?

Wouldn't one usually provide the `domain` (i.e. in the hkdf step _info_) in the _expand_ step?

Reply:
Yeah, you are right. We should have used `from_prk` and `expand` using the provided domain.